### ISSUE TITLE
Bayesian Optimisation - Test Set Contamination

### ISSUE THREAD

**Issue by lpcinelli** [emtiyaz/vadam#3]
According to Yarin Gal's repo [DropoutUncertaintyExps](https://github.com/yaringal/DropoutUncertaintyExps/), there was test set contamination due to the bayesian optmisation of hypers:
> Update (2018) We replaced the Bayesian optimisation implementation (which was used to find hypers) with a grid-search over the hypers. This is following feedback from @capybaralet who spotted test-set contamination (some train-set points, used to tune hypers which were shared across all splits, were used as test-set points in later splits). The new implementation iterates over the 20 splits, and for each train-test split it creates a new train-val split to tune hypers. These hypers are discarded between different train-test splits.

Yet, it seems the UCI regression experiments here use the same splits as in the [DropoutUncertaintyExps](https://github.com/yaringal/DropoutUncertaintyExps/) repo. Was the bug fixed or is it still there and the reason for doing so here was the reviewers' request?

**Comment by emtiyaz** [comment]
@didriknielsen could you follow up on this?

**Comment by didriknielsen** [comment]
The splits themselves were not the issue. These were not changed in the newest commit, only renamed. See e.g. [here](https://github.com/yaringal/DropoutUncertaintyExps/commit/6eb4497628d12b0f300f4b4f6bdc386bebad565c#diff-5568a97bdba072a7ab1278a7e57cd451).

As far as I understand it, the issue was that the hypers were not discarded between different train-test splits, leading to the contamination.

This is not the case in our experiments as for each of the 20 train-test splits, we run a separate BO with 5-fold CV over the train set which is used to select the hypers for that split. Thus, we obtain 20 hypers, one for each split (which is used for testing on that split). 

In the original MC-dropout code (see e.g. [here](https://github.com/yaringal/DropoutUncertaintyExps/blob/a6259f1db8f5d3e2d743f88ecbde425a07b12445/bostonHousing/net/experiment_BO.py)), a single BO was run over the first 5 / 20 of the train-test splits. The results for each hyperparameter were averaged over the 5 splits to select a single hyperparameter for the full dataset (instead of 20, one for each split).

**Comment by lpcinelli** [comment]
Thank you for pointing out the original code and clarifying the implementation problem. Also, thanks for making your code available, it's been really helpful !
